File: android_cursor_grok/data/crypto/src/main/kotlin/com/transcripto/crypto/CaesarCipher.kt
package com.transcripto.crypto

import com.transcripto.domain.CipherMethod

class CaesarCipher : CipherMethod {
    override fun encrypt(input: String, params: Map<String, Any>): String {
        val shift = (params["shift"] as? Int ?: 0) + getSaltShift(params["salt"] as? String ?: "", true)
        return input.map { char -> shiftChar(char, shift) }.joinToString("")
    }

    override fun decrypt(input: String, params: Map<String, Any>): String {
        val shift = (params["shift"] as? Int ?: 0) + getSaltShift(params["salt"] as? String ?: "", false)
        return input.map { char -> shiftChar(char, shift) }.joinToString("")
    }

    private fun shiftChar(c: Char, shift: Int): Char {
        val effectiveShift = shift % 26
        return when {
            c.isUpperCase() -> ((c - 'A' + effectiveShift + 26) % 26 + 'A'.code).toChar()
            c.isLowerCase() -> ((c - 'a' + effectiveShift + 26) % 26 + 'a'.code).toChar()
            else -> c
        }
    }

    private fun getSaltShift(salt: String, isEncrypt: Boolean): Int {
        if (salt.isEmpty()) return 0
        val hash = salt.hashCode() % 26
        return if (isEncrypt) hash else -hash
    }
}

File: android_flutter/transcripto/packages/data_crypto/lib/src/caesar_cipher.dart

import 'package:domain/domain.dart';
import 'package:core_common/core_common.dart';

class CaesarCipher implements CipherMethod {
  int _normalizeShift(int shift) => ((shift % 26) + 26) % 26;

  @override
  String encrypt(String input, CipherParams params) {
    final shift = _effectiveShift(params);
    final out = input.split('').map((c) => _shiftChar(c, shift)).join();
    return out;
  }

  @override
  String decrypt(String input, CipherParams params) {
    // invert encryption shift (base + saltAdj)
    final encShift = _effectiveShift(params);
    return input.split('').map((c) => _shiftChar(c, -encShift)).join();
  }

  int _effectiveShift(CipherParams p) {
    final base = p.shift ?? 0;
    var saltAdj = 0;
    if (p.useSalt && p.salt != null) {
      saltAdj = simpleHashBytes(p.salt!) % 26;
    }
    return _normalizeShift(base + saltAdj);
  }

  String _shiftChar(String c, int shift) {
    if (c.isEmpty) return c;
    final code = c.codeUnitAt(0);
    if (code >= 65 && code <= 90) {
      // A-Z
      final a = 65;
      final pos = code - a;
      final npos = (pos + shift) % 26;
      return String.fromCharCode(a + (npos < 0 ? npos + 26 : npos));
    } else if (code >= 97 && code <= 122) {
      // a-z
      final a = 97;
      final pos = code - a;
      final npos = (pos + shift) % 26;
      return String.fromCharCode(a + (npos < 0 ? npos + 26 : npos));
    }
    return c; // non alpha unchanged
  }
}
